### PR TITLE
Add support to draw base shapes

### PR DESCRIPTION
**Pull request by VDykyi**

### CODE REVIEW

File: src/VAC/Global.cpp [hunk 1]
```diff
@@ -653,6 +653,33 @@ void Global::setToolMode(Global::ToolMode mode)
...
+        separatorSketch2_->setVisible(true);
+        break;
+    case DRAW_CIRCLE:
+        actionPlanarMapMode_->setVisible(true);
+        actionSnapMode_->setVisible(true);
+        actionUseTabletPressure_->setVisible(true);
+        actionSnapThreshold_->setVisible(true);
+        actionEdgeWidth_->setVisible(true);
+        separatorSketch1_->setVisible(true);
+        separatorSketch2_->setVisible(true);
```
Comment (marcus-sonestedt) [inline src/VAC/Global.cpp]: All cases seem to run the same code? Maybe show that via `case x: case y: case z: {code }` ?

File: src/VAC/Scene.cpp [hunk 2]
```diff
@@ -51,7 +51,7 @@ Scene * Scene::createDefaultScene()
 {
     Scene * res = new Scene();
     Layer * layer = res->createLayer(tr("Layer 1"));
-    layer->background()->setColor(Qt::white);
+    layer->background()->setColor(QColor::fromRgb(240, 241, 242));
```
Comment (marcus-sonestedt) [inline src/VAC/Scene.cpp:55]: It'd be nice to not hard-code Cellink's colors into VPaint. Similar with glClearColor() below.

Not sure how to best solve it though. Easiest is probably to move all colors to a separate header and use #ifdef CELLINK_VPAINT_STYLE to choose this or that, and then #define that based on some qmake-property in vpaint.pro

File: src/VAC/View.cpp [hunk 3]
```diff
@@ -706,6 +724,68 @@ void View::PMRPressEvent(int action, double x, double y)
...
+        {
+          if(mouse_isTablet_ &&  global()->useTabletPressure())
+            w *= 2 * mouse_tabletPressure_; // 2 so that a half-pressure would get the default width
+        }
+        vac_->beginSketchEdge(xScene,yScene, w, interactiveTime());
+        //vac_->continueSketchEdge(xScene +100,yScene, w);
+        //vac_->endSketchEdge();
+
+        //emit allViewsNeedToUpdatePicking();
+        //updateHoveredObject(mouse_Event_X_, mouse_Event_Y_);
```
Comment (marcus-sonestedt) [inline src/VAC/View.cpp]: please either remove commented code, or add new comment line explaining why it should remain.

File: src/VAC/View.cpp [hunk 4]
```diff
@@ -706,6 +724,68 @@ void View::PMRPressEvent(int action, double x, double y)
...
+
+        double w = global()->settings().edgeWidth();
+        bool debug = false;
+        if(!debug)
+        {
+          if(mouse_isTablet_ &&  global()->useTabletPressure())
+            w *= 2 * mouse_tabletPressure_; // 2 so that a half-pressure would get the default width
+        }
+        //vac_->beginSketchEdge(shapeStartX, shapeStartY, w, interactiveTime());
+        //emit allViewsNeedToUpdate();
```
Comment (marcus-sonestedt) [inline src/VAC/View.cpp]: previous two cases includes these two lines.  Should we keep or remove them?

File: src/VAC/View.cpp [hunk 5]
```diff
@@ -955,6 +1035,106 @@ void View::PMRReleaseEvent(int action, double x, double y)
...
+
+      double radius = sqrt((xScene - shapeStartX) * (xScene - shapeStartX) + (yScene - shapeStartY) * (yScene - shapeStartY)) / 2;
+      double centerX = (xScene + shapeStartX) / 2;
+      double centerY = (yScene + shapeStartY) / 2;
+
+      double startX = (xScene + shapeStartX) / 2 + radius;
+      double startY = (yScene + shapeStartY) / 2;
+      vac_->beginSketchEdge(startX, startY, w, interactiveTime());
+
+      for (double deg = 0; deg <= M_PI * 2; deg += 0.1)
```
Comment (marcus-sonestedt) [inline src/VAC/View.cpp]: could we compute the increment based on the size of the circle? .. so that we step 1mm?

File: src/VAC/View.cpp [hunk 6]
```diff
@@ -1678,6 +1859,11 @@ QImage View::drawToImage(Time t, double x, double y, double w, double h, int IMG
     return res;
 }
 
+void View::drawLine()
+{
+
```
Comment (marcus-sonestedt) [inline src/VAC/View.cpp]: intentionally empty or mistake? Add a comment to explain what's needed here, or just // empty because xxx...

**Review by marcus-sonestedt — APPROVED** [review]
Good start. I spotted some things that might be correct but confused me a bit.

File: src/VAC/Global.cpp (hunk 1 above)
Comment (VDykyi) [reply src/VAC/Global.cpp]: Done

File: src/VAC/Scene.cpp (hunk 2 above)
Comment (VDykyi) [reply src/VAC/Scene.cpp:55]: Done

File: src/VAC/View.cpp (hunk 3 above)
Comment (VDykyi) [reply src/VAC/View.cpp]: Done

File: src/VAC/View.cpp (hunk 4 above)
Comment (VDykyi) [reply src/VAC/View.cpp]: Done

File: src/VAC/View.cpp (hunk 5 above)
Comment (VDykyi) [reply src/VAC/View.cpp]: I agree. But in this implementation every call of  vac_->continueSketchEdge(newX, newY, w); performs rounding and quality of circle doesn't depend of it radius. I just reduce the degree increment to 0.01 and circles of all size will look fine.

But in the next iteration I plan to change the drawing circle algorithm (like polygons) and will use dependency between circle size and number of the polygon lines.

Maybe I will merge so far as is yet?
Do you aprove?

File: src/VAC/View.cpp (hunk 6 above)
Comment (VDykyi) [reply src/VAC/View.cpp]: Done

File: src/VAC/View.cpp (hunk 5 above)
Comment (marcus-sonestedt) [reply src/VAC/View.cpp]: Understood. Makes sense. I approve, go ahead and merge.

We probably need to have some "minimal travel distance" optimization when we create gcode anyway.

**Comment by marcus-sonestedt** [comment]
Btw, just mark conversation as "resolved" when you're done with them.

If purpose and fix is clear, there is no need to comment and wait for me to resolve. :)

**Pull request merged**